File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestOnes = function(nums, k) {
    let n = nums.length;
    let zero = 0;
    let maxlen = 0;
    let one = 0;

    for(let i=0; i<n; i++){
        if(nums[i] == 0){
            zero++;
        }
        while(zero > k){
            if(nums[one] == 0){
                zero--;
            }
            one++;
        }
        maxlen = Math.max(maxlen , i-one+1);
    }
    return maxlen;
    
};

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int left = 0;
        int z= 0;
        int max= 0;
        int count = 0;

        for(int right=0; right<nums.length; right++){
            if(nums[right] == 0){
                z++;
                while(z>k){
                    if(nums[left] == 0){
                        z--;
                    }
                    left++;
                }
            }
            int len = right - left + 1;
            max = Math.max(max , len);
        }
        return max;
    }
}